app/src/Sonata/Admin/UserAdmin.php

```diff
@@ -220,6 +222,21 @@ public function getActions(): array
             'addRoles' => (new AdminAction('addRoles', true))
                 ->setController(AddRolesAminAction::class)
                 ->setBatchController(AddRolesAminAction::class),
-        ];
+        ] + iterator_to_array($this->prepareSetPreferredLocaleActions());
+    }
+
+    private function prepareSetPreferredLocaleActions(): iterable
+    {
+        foreach (['en', 'fr'] as $locale) {
+            yield $name = \sprintf('setPreferredLocale%s', ucfirst($locale)) => (new AdminAction($name, true))
+                ->setLabel(strtoupper($locale))
+                ->setIcon('fa fa-language')
+                ->setController(SetPreferredLocaleAction::class)
+                ->setRoutePattern(
+                    \sprintf('%s/preferred-locale/{_locale}', $this->getRouterIdParameter())
+                )
+                ->addRouteParam('_locale', $locale)
```

Comment:
Since the route is base on the action name not sure it will work, the last one will override the previous one.

The param must be added dynamically at runtime not at "preparation".